<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet.

Use the links below to read the data files using the Pandas library.

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [1]:
!pip3 install ipython-sql
!pip install sqlalchemy
import sqlalchemy
import sqlite3

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 17.1 MB/s eta 0:00:00


Load the SQL magic module


In [3]:
%load_ext sql

Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [5]:
import pandas as pd

cd_df=pd.read_csv("Census_Data_-_Selected_socioeconomic_indicators_in_Chicago__2008___2012_20240311.csv",index_col=0)
cd_df.describe()

ccd_df=pd.read_csv("ChicagoCrimeData.csv",index_col=0)
ccd_df.describe()

cps_df=pd.read_csv("Chicago_Public_Schools_-_Progress_Report_Cards__2011-2012__20240311.csv",index_col=0)
cps_df.describe()

ZIP Code  Safety Score  Environment Score  Instruction Score  \
count    566.000000    513.000000         513.000000         513.000000   
mean   60629.727915     49.504873          47.766082          48.288499   
std       20.248024     20.110837          16.215584          17.417176   
min    60605.000000      1.000000           1.000000           1.000000   
25%    60618.000000     35.000000          37.000000          37.000000   
50%    60625.000000     48.000000          47.000000          47.000000   
75%    60639.000000     61.000000          58.000000          59.000000   
max    60827.000000     99.000000          99.000000          99.000000   

       Average Student Attendance  Rate of Misconducts (per 100 students)   \
count                  565.000000                               566.000000   
mean                    92.313805                                21.085866   
std                      5.993053                                27.543835   
min                     57.900000                                 0.000000   
25%                     91.900000                                 4.525000   
50%                     94.400000                                12.250000   
75%                     95.600000                                26.975000   
max                     98.400000                               251.600000   

       Average Teacher Attendance  \
count                  566.000000   
mean                    94.378445   
std                     11.353180   
min                      0.000000   
25%                     95.100000   
50%                     95.800000   
75%                     96.400000   
max                     98.500000   

       Individualized Education Program Compliance Rate   \
count                                         566.000000   
mean                                           98.915371   
std                                             2.311631   
min                                            72.700000   
25%                                            98.600000   
50%                                           100.000000   
75%                                           100.000000   
max                                           100.000000   

       ISAT Exceeding Math %  ISAT Exceeding Reading %   ...  \
count             476.000000                 476.000000  ...   
mean               20.356303                  15.667017  ...   
std                16.850870                  15.208361  ...   
min                 0.000000                   0.000000  ...   
25%                 9.175000                   6.100000  ...   
50%                16.000000                  10.950000  ...   
75%                24.900000                  19.525000  ...   
max               100.000000                 100.000000  ...   

       College Enrollment (number of students)   General Services Route   \
count                                566.000000               566.000000   
mean                                 626.053004                39.084806   
std                                  448.495105                 6.165552   
min                                   21.000000                29.000000   
25%                                  342.250000                34.000000   
50%                                  504.000000                39.000000   
75%                                  790.500000                45.000000   
max                                 4368.000000                49.000000   

         RCDTS Code  X_COORDINATE  Y_COORDINATE    Latitude   Longitude  \
count  5.660000e+02  5.660000e+02  5.660000e+02  566.000000  566.000000   
mean   1.500000e+14  1.163016e+06  1.885663e+06   41.841890  -87.677323   
std    0.000000e+00  1.605930e+04  3.285688e+04    0.090361    0.058390   
min    1.500000e+14  1.118114e+06  1.817242e+06   41.653674  -87.841052   
25%    1.500000e+14  1.152166e+06  1.860248e+06   41.771866  -87.717152   
50%    1.500000e+14  1.163553e+06  1.886889e+

Establish a connection between SQL magic module and the database `FinalDB.db`


In [6]:
%sql sqlite:///FinalDB.db

You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [14]:
%sql SELECT COUNT(ID) FROM CHICAGO_CRIME_DATA

 * sqlite:///FinalDB.db
Done.


[(533,)]

### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [161]:
%%sql
SELECT COMMUNITYAREANAME as "Name", CommunityAreaNumber AS "Number"
FROM CENSUS_DATA
WHERE PERCAPITAINCOME < 11000

 * sqlite:///FinalDB.db
Done.


[('West Garfield Park', 26),
 ('South Lawndale', 30),
 ('Fuller Park', 37),
 ('Riverdale', 54)]

### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis)


In [18]:
%sql SELECT CASE_NUMBER FROM CHICAGO_CRIME_DATA WHERE DESCRIPTION LIKE '%MINOR%'

 * sqlite:///FinalDB.db
Done.


[('HL266884',), ('HK238408',)]

### Problem 4

##### List all kidnapping crimes involving a child?


In [21]:
%%sql
SELECT * FROM CHICAGO_CRIME_DATA
WHERE PRIMARY_TYPE = "KIDNAPPING" AND DESCRIPTION LIKE '%CHILD%'

 * sqlite:///FinalDB.db
Done.


[(5276766, 'HN144152', '2007-01-26', '050XX W VAN BUREN ST', 1792, 'KIDNAPPING', 'CHILD ABDUCTION/STRANGER', 'STREET', 'FALSE', 'FALSE', 1533, 15, 29, 25, 20, 1143050, 1897546, 2007, 41.87490841, -87.75024931, '(41.874908413, -87.750249307)')]

### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


In [43]:
%%sql
SELECT DISTINCT (PRIMARY_TYPE), DESCRIPTION FROM CHICAGO_CRIME_DATA
WHERE LOCATION_DESCRIPTION LIKE '%SCHOOL%'

 * sqlite:///FinalDB.db
Done.


[('BATTERY', 'SIMPLE'),
 ('BATTERY', 'PRO EMP HANDS NO/MIN INJURY'),
 ('CRIMINAL DAMAGE', 'TO VEHICLE'),
 ('NARCOTICS', 'POSS: HEROIN(WHITE)'),
 ('NARCOTICS', 'MANU/DEL:CANNABIS 10GM OR LESS'),
 ('ASSAULT', 'PRO EMP HANDS NO/MIN INJURY'),
 ('CRIMINAL TRESPASS', 'TO LAND'),
 ('PUBLIC PEACE VIOLATION', 'BOMB THREAT')]

### Problem 6

##### List the type of schools along with the average safety score for each type.


In [136]:
%%sql SELECT ElementaryMiddleorHighSchool, AVG(SafetyScore) Average_Safety_Scores
FROM CHICAGO_PUBLIC_SCHOOLS
GROUP BY "ElementaryMiddleorHighSchool"

 * sqlite:///FinalDB.db
Done.


[('ES', 49.52038369304557), ('HS', 49.62352941176471), ('MS', 48.0)]

### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [137]:
%%sql SELECT CommunityAreaName, CommunityAreaNumber
FROM CENSUS_DATA
ORDER BY PERCENTHOUSEHOLDSBELOWPOVERTY DESC LIMIT 5

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 54),
 ('Fuller Park', 37),
 ('Englewood', 68),
 ('North Lawndale', 29),
 ('East Garfield Park', 27)]

### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [163]:
%%sql
SELECT COMMUNITY_AREA_NUMBER
FROM CHICAGO_CRIME_DATA
GROUP BY COMMUNITY_AREA_NUMBER
ORDER BY COUNT(COMMUNITY_AREA_NUMBER) DESC
LIMIT 1

 * sqlite:///FinalDB.db
Done.


[(25,)]


Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [126]:
%sql \
SELECT COMMUNITYAREANAME AS "NAME", HARDSHIPINDEX AS "INDEX" FROM CENSUS_DATA \
WHERE HARDSHIPINDEX = (SELECT MAX(HARDSHIPINDEX) FROM CENSUS_DATA)

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 98)]

### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [158]:
%%sql
SELECT COMMUNITYAREANAME AS "Name"
FROM CENSUS_DATA
WHERE CommunityAreaNumber =
    (SELECT COMMUNITY_AREA_NUMBER
    FROM CHICAGO_CRIME_DATA
    GROUP BY COMMUNITY_AREA_NUMBER
    ORDER BY COUNT(COMMUNITY_AREA_NUMBER) DESC
    LIMIT 1
    )

 * sqlite:///FinalDB.db
Done.


[('Austin',)]

## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>

## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |

## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
